server.js

```diff
@@ -183,6 +183,7 @@ app.prepare()
      */
     express.get('/webcam/stream', (req, res) => {
     // Proxy MJPEG stream from darknet to avoid freezing issues
+      if(!Opendatacam.isListeningToYOLO && mjpgProxy !== null) mjpgProxy = null; //This clears the mjpgProxy object in case YOLO is restarting otherwise a grey screen appears in the UI.
```

Comment:
I feel like this only partially fixed the error. 

E.G. what happens if the restart of Yolo finishes before someone refreshes the WebUI. In this case `isListeningToYolo` will again the true and the MJPEG Proxy will not be refreshed. 

In general I wonder if the MJPEG Proxy can detect and fix broken streams on its own. Without having to force restart it.

Reply:
Well, I am not certain that would happen - but on different hardware maybe?? When I was troubleshooting this, I had debug statements every line so I could trace the exact order (doing breakpoints proved to be ineffective due to the asynchronous nature of some of the operations).  Anyway, consistently it hit that endpoint prior to loading UI, so I do feel it would work no matter what, but I only have the 2 jetsons to test on and I don't have a linux desktop (with Nvidia gpu).

I am not well versed on forever monitor, but that might be a possible solution to monitor the proxy.  Thoughts?

Reply:
From the point of MJPEG Proxy, all that happens is that the MJPEG Stream gets disconnected. It does not need to know that this is because of Darknet crashing, could just be a Network fault.

Quick tests show that when I crash Darknet manually, the `close` event inside MJPEG Proxy gets raised. However there is no handling of the event. Therefore instead of monitoring Darknet externally I would like to limit the dependencies between ODC and Darknet as much as possible and handle the MJPEG reconnect inside the Proxy.

Reply:
Ok, that makes sense. Let me look into hooking into the close event.

Reply:
So regarding hooking into the close event, the issue is the `mjpegProxy` variable.  So here is a thought - again I am completely new to ODC but from what my tests are showing, this seems to be a way forward - we move the declaration of `mjpgProxy` into `Opendatacam.js` so we would reference `Opendatacam.mjpgProxy` wherever its needed.  We import that file into `mjpegProxy.js`, then in the `onClose` event we just need to set `Opendatacam.mjpgProxy` to `null`.  We can get rid of this line `if(!Opendatacam.isListeningToYOLO && mjpgProxy !== null) mjpgProxy = null;` as the actual call to the `/webcam/stream` endpoint would then re-establish the connection.  This should make it more agnostic and aware.  Thoughts on that?

Reply:
I don't like this for two reasons

1. Recreating `mjpgProxy` would close all browser clients, and require them to reconnect/refresh the page. This does not seem like a big deal in your case as the UI gets refreshed anyway as Darknet is intentionally restarted, but it will make a difference to situations where Darknet crashes in the back.
2. In the long term I would like to make ODC independent of Darknet to potentially allow plugging in different Engines such as TensorFlow or Deepstream. These engines might not require a MJPEG Proxy. Therefore I would love to keep `Opendatacam.js` independent of Video Streaming and implementation details of the neural network. See for example https://github.com/opendatacam/opendatacam/issues/342 or https://github.com/opendatacam/opendatacam/issues/112 or https://github.com/opendatacam/opendatacam/pull/462